### PR TITLE
Remove transformer callback

### PR DESCRIPTION
**Pull request by towhans**

### CODE REVIEW

File: lib/kafka_gen_stage/consumer_logic.ex [hunk 1]
```diff
@@ -28,71 +27,76 @@ defmodule KafkaGenStage.ConsumerLogic do
...
-          prepare_dispatch(queue, new_demand, Enum.reverse(to_send), to_ack, transformer, new_state)
-
-        new_to_send = transform_bulk(to_send, bulk_transformer)
         final_demand = lower_demand(demand, new_to_send)
-        {new_to_send, to_ack, final_demand, queue, new_transformer_state}
+        {new_to_send, to_ack, final_demand, queue}
 
       {:empty, queue} ->
-        {[], :no_ack, demand, queue, state}
+        # proc se ne-ackne, kdyz je queue prazdna? proc se na-ackne predchozi zprava?
```
Comment (Coffei) [inline lib/kafka_gen_stage/consumer_logic.ex]: Is this something you'd want to leave in the code? If so, please, translate to english.

File: lib/kafka_gen_stage/consumer_logic.ex [hunk 2]
```diff
@@ -28,71 +27,76 @@ defmodule KafkaGenStage.ConsumerLogic do
...
-          transformer_state :: term,
-          bulk_transformer :: ([msg_tuple()] -> [msg_tuple()])
+          bulk_transformer :: ([msg_tuple()] -> [msg_tuple()]),
+          high_wm :: non_neg_integer()
         ) ::
           dispatch()
-  def prepare_dispatch(queue, 0, _transformer, state, _bulk_transformer) do
-    {[], :no_ack, 0, queue, state}
+  def prepare_dispatch(queue, 0, bulk_transformer, high_wm) do
+    {transform_bulk([], bulk_transformer, high_wm), :no_ack, 0, queue}
```
Comment (Coffei) [inline lib/kafka_gen_stage/consumer_logic.ex]: Should we send something when the demand is `0`?

File: lib/kafka_gen_stage/consumer_logic.ex (hunk 1 above)
Comment (Coffei) [reply lib/kafka_gen_stage/consumer_logic.ex]: Not 100% sure, but it seems that message is acked when it's moved from `queue` to `to_send`. Here there's nothing in queue, therefore, everything that was in queue before was already acked (it was already moved from `queue` to `to_send`). So the present logic seems correct to me. Do you have any indication it's not correct?

**Review by Coffei — COMMENTED** [review]
Overall the changes look good!

File: lib/kafka_gen_stage/consumer_logic.ex (hunk 2 above)
Comment (towhans) [reply lib/kafka_gen_stage/consumer_logic.ex]: No. You're right.

File: lib/kafka_gen_stage/consumer_logic.ex (hunk 1 above)
Comment (towhans) [reply lib/kafka_gen_stage/consumer_logic.ex]: It's how you say it is. Still I added transform_bulk to the following line which was not there before. Please check that too.

Comment (Coffei) [reply lib/kafka_gen_stage/consumer_logic.ex]: That seems OK.

**Review by Coffei — APPROVED** [review]

File: lib/kafka_gen_stage/consumer.ex [hunk 3]
```diff
@@ -154,15 +141,14 @@ defmodule KafkaGenStage.Consumer do
           consumer: pid(),
           consumer_ref: reference(),
           queue: :queue.queue(),
+          high_wm_offset: non_neg_integer() | atom(),
```
Comment (liskin) [inline lib/kafka_gen_stage/consumer.ex:144]: Why not `:unknown` or perhaps better still `nil`? Can we get an atom from brod?

File: lib/kafka_gen_stage/consumer_logic.ex [hunk 4]
```diff
@@ -28,71 +27,78 @@ defmodule KafkaGenStage.ConsumerLogic do
   @spec prepare_dispatch(
           buffered_msgs :: :queue.queue(),
           buffered_demand :: non_neg_integer(),
-          transformer :: (msg_tuple(), term() -> {term(), term()}),
-          transformer_state :: term,
-          bulk_transformer :: ([msg_tuple()] -> [msg_tuple()])
+          bulk_transformer :: ([msg_tuple()] -> [msg_tuple()]),
+          high_wm :: non_neg_integer()
```
Comment (liskin) [inline lib/kafka_gen_stage/consumer_logic.ex:31]: I think this type is incorrect. If a consumer asks for demand sooner than we get anything from kafka, then this function is called with `:unknown`. I'm not entirely sure how likely that is to happen, so please correct me if I'm wrong.

File: lib/kafka_gen_stage/consumer_logic.ex [hunk 5]
```diff
@@ -28,71 +27,78 @@ defmodule KafkaGenStage.ConsumerLogic do
...
+          new_to_send =
+            to_send
+            |> Enum.reverse()
+            |> transform_bulk(bulk_transformer, high_wm)
+
+          {new_to_send, to_ack, queue}
+
+        {:empty, queue} ->
+          {transform_bulk([], bulk_transformer, high_wm), :no_ack, queue}
+      end
```
Comment (liskin) [inline lib/kafka_gen_stage/consumer_logic.ex:60]: There seems to be some unnecessary code duplication in prepare_dispatch/4 and prepare_dispatch/5. Why don't we unconditionally call prepare_dispatch/5 here? Something like this:

```
{to_send, to_ack, _demand, queue} = prepare_dispatch(queue, demand, [], :no_ack, high_wm)
to_send = … |> transform_bulk(…)
demand = lower_demand(demand, to_send)
{to_send, to_ack, demand, queue}
```

And we can also drop the unused `_demand` from prepare_dispatch/5 return value.

File: lib/kafka_gen_stage/consumer_logic.ex (hunk 4 above)
Comment (liskin) [reply lib/kafka_gen_stage/consumer_logic.ex:31]: (You do seem to handle this case in transform_bulk/3 so I guess it does happen.)

Comment (towhans) [reply lib/kafka_gen_stage/consumer_logic.ex:31]: Exactly, it does not happen.

Comment (liskin) [reply lib/kafka_gen_stage/consumer_logic.ex:31]: Why do you handle it then?

File: lib/kafka_gen_stage/consumer_logic.ex (hunk 5 above)
Comment (towhans) [reply lib/kafka_gen_stage/consumer_logic.ex:60]: I agree. I will do it.

File: lib/kafka_gen_stage/consumer.ex (hunk 3 above)
Comment (towhans) [reply lib/kafka_gen_stage/consumer.ex:144]: OK, will use `nil`

File: lib/kafka_gen_stage/consumer_logic.ex (hunk 4 above)
Comment (towhans) [reply lib/kafka_gen_stage/consumer_logic.ex:31]: It does happen exactly in the case that is handled in `transform_bulk/3` by not calling the bulk_transformer fun.

**Pull request merged**